rfc: snpguest first version

DGonzalezVillal opened the pull request:
Initial version of the snpguest utility tool. Can request report, extended report, certificate chain from kds and VCEK from kds. Can also verify root of trust.

tylerfanelli commented on src/request.rs:
```diff
@@ -0,0 +1,430 @@
+    //Request extended report function
+    pub fn get_extended_report(args: Args) -> Result<()> {
+        
+        // Open sev firmware device
```
```suggestion
```

tylerfanelli commented on src/request.rs:
```diff
@@ -0,0 +1,430 @@
+            .read(&mut request_data)
+            .context("Could not read report request file.")?;
+        
+        // Create snp request structure
```
```suggestion
```

tylerfanelli commented on src/request.rs:
```diff
@@ -0,0 +1,430 @@
+        // Create snp request structure
+        let snp_data: SnpReportReq = SnpReportReq::new(Some(request_data), args.vmpl);
+
+        // Request extended report
```
```suggestion
```

tylerfanelli commented on src/request.rs:
```diff
@@ -0,0 +1,430 @@
+}
+
+// Module to request extended-report
+mod extended_report {
```
Instead of `get_report` and `get_extended_report`, can `get_report` just include a flag `--extended` to include the certs?

larrydewey replied:
Seems reasonable enough to me.

tylerfanelli reviewed (changes requested):
Thanks @DGonzalezVillal, I've left some comments.

Ultimately, I think this PR can hold off until we release the initial `virtee/snp` crate (whether that is included in the `virtee/sev` crate or by itself). This crate will allow these commands to be implemented much simpler, especially when concerning certificates.

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,72 @@ GLOBAL OPTIONS
+    usage: snpguest request report --message_version $MESSAGE_VERSION_NUMBER --request_file $REQUEST_FILE_PATH --vmpl $VMPL --att_report_path $ATTESTATION_REPORT_PATH
+    
+    Requests an attestation report from the host. Will write the contents of the attestation report in binary format into the specified report path. 
+    User needs to pass 64 bytes of data in any file format into request_file in order to be able to request the attestation report.
```
Although it isn't required, it is strongly suggested. The unique data from the user provides an additional means of verifying the report hasn't been tampered with. Perhaps if they don't specify one, we could create a pseudo-random buffer and print it out for the user to look for it in the report?

tylerfanelli replied:
Yes, that can be a viable option.

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,72 @@ GLOBAL OPTIONS
+    usage: snpguest request extended-report --message_version $MESSAGE_VERSION_NUMBER --request_file $REQUEST_FILE_PATH --vmpl $VMPL --att_report_path $ATTESTATION_REPORT_PATH --ark_path $ARK_PATH --ask_path $ASK_PATH --vcek_path $vcek_path
+
+    Requests an extended attestation report from the host. Will write the contents of the attestation report in binary format into specified report path. 
+    User needs to pass 64 bytes of data in any file format into request_file in order to be able to request the attestation report.
```
Here, I believe the request file is the means of passing the 64-byte value. Depending on how we take the application, it *could* be required.

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,72 @@ GLOBAL OPTIONS
+    usage: snpguest request certificate-chain --processor_type $PROCESSOR_TYPE --ark_path $ARK_PATH  --ask_path $ASK_PATH
+
+    Requests the CA certificate chain from the KDS by creating the URL.
+    The user should specify their host processory type using the processory_type flag, it will default to "Milan".
```
Agreed. It would be a good idea to have this a bit more dynamic.

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,72 @@ GLOBAL OPTIONS
+    usage: snpguest request vcek --processor_type $PROCESSOR_TYPE --vcek_path $VCEK_PATH  --att_report_path $ATTESTATION_REPORT_PATH
+
+    Requests the vcek certificate from the KDS.
+    The user should specify their host processory type using the processory_type flag, it will default to "Milan".
```
+1

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,72 @@ GLOBAL OPTIONS
+    Requests the vcek certificate from the KDS.
+    The user should specify their host processory type using the processory_type flag, it will default to "Milan".
+    The user needs to provide a path to an attestation report in order to request the vcek.
+    If no path is provided for attestation report, the program will look for an attestation report in the local directory by default ("./attestation_report.bin").
```
This would be pretty helpful. I am trying to think through a scenario where a user may __not__ want this to occur. What is the VM / container is running in an isolated network?

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,72 @@ GLOBAL OPTIONS
+    -h, --help show a help message
+
+*snpguest guest verify-root-of-trust*::
+    usage: snpguest verify-root-of-trust --ark_path $ARK_PATH --ask_path $ASK_PATH --vcek_path $vcek_path
```
True. I think this could be re-worded to verify certificate chain, or add in the report.

larrydewey commented on src/request.rs:
```diff
@@ -0,0 +1,430 @@
+            default_value = "0",
+            help = "VMPL level the Guest is running on."
+        )]
+        pub vmpl: u32,
```
Technically the API doesn't required a value, so this can be an `Option`.

DGonzalezVillal replied:
Will make it option, will default to 1 if no value is provided

DGonzalezVillal commented:
Restructuring of the project has now been done in this latest command. I also addressed all the comments and the project is using the latest sev library. I know there are a couple of things that could be done in a quicker way (less code), but I also want to make sure I'm outputting as much information as possible to the user.

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    -r, --random  generate 64 random bytes of data to pass for the report request
+
+*snpguest fetch ca*::
+    usage: snpguest fetch ca [-p,--processor] $PROCESSOR_TYPE [-c, --certs] $CERTS_PATH
```
```suggestion
    usage: snpguest fetch ca $PROCESSOR_TYPE $CERTS_PATH
```

Usually, with flags such as `-p, --processor` and `-c, --certs` it indicates that they are optional. In this command, they are not. We can remove these.

DGonzalezVillal replied:
I understand. I removed the processor type flag. Shouldn't certs be optional if I have a default directory where the program looks for certs? It currently looks for the certs in "./certs" if no path is provided.

tylerfanelli replied:
@DGonzalezVillal You can require it altogether

DGonzalezVillal replied:
@tylerfanelli understood. If the provided directory already exists do we want to create an error or create it for them?

tylerfanelli replied:
@DGonzalezVillal 

If directory exists already ---> use it and overwrite any cert files.
If directory doesn't exist already ---> create it and fill it with the certs

Sounds like this would be a nice way to do things?

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    usage: snpguest fetch ca [-p,--processor] $PROCESSOR_TYPE [-c, --certs] $CERTS_PATH
+
+    Requests the certificate authority chain (ARK & ASK) from the KDS.
+    The user must specify their host processor type using the --processor flag.
```
```suggestion
    The user must specify their host processor type.
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    -h, --help show a help message
+
+*snpguest fetch vcek*::
+    usage: snpguest fetch vcek [-p, --processor] $PROCESSOR_TYPE [-c, --certs] $CERTS_PATH  [-a, --att-report] $ATTESTATION_REPORT_PATH
```
```suggestion
    usage: snpguest fetch vcek $PROCESSOR_TYPE $CERTS_PATH
```

Same here, remove the flag indicators. Also, why not just request the attestation report from the PSP?

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+
+    Requests the VCEK certificate from the KDS.
+    The user must specify their host processor type using the processor flag.
+    The user needs to provide the path to an existing attestation report in order to request the VCEK.
+    If no path is provided for the attestation report, the program will look for an attestation report in the local directory by default ("./attestation_report.bin").
```
```suggestion
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    options:
+    -h, --help show a help message
+
+*snpguest guest verify certificate-chain*::
```
```suggestion
*snpguest guest verify certs*::
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    -h, --help show a help message
+
+*snpguest guest verify certificate-chain*::
+    usage: snpguest verify certificate-chain [-c, --certs] $CERTS_PATH
```
```suggestion
    usage: snpguest verify certs $CERTS_PATH
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+
+    Verifies that the provided certificate chain has been properly signed by each certificate.
+    The user needs to provide a directory were all 3 of the certificates are stored (ARK,ASK and VCEK).
+    If no path is provided for the certificates, the program will look for them in the local cert directory ("./certs")
```
Make the path required, instead of searching `./certs`?

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    -h, --help show a help message
+
+*snpguest verify tcb*::
+    usage: snpguest verify tcb [-c, --certs] $CERTS_PATH [-a, --att-report] $ATTESTATION_REPORT_PATH
```
```suggestion
    usage: snpguest verify tcb $VCEK
```

It's only the VCEK we're looking for here, so just give that file. Also fetch the attestation report from the PSP?

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    usage: snpguest verify tcb [-c, --certs] $CERTS_PATH [-a, --att-report] $ATTESTATION_REPORT_PATH
+
+    Verifies the trusted compute base using the provided attestation report and VCEK certificate
+    The user needs to provide a path to the attestation report and to the directory where the VCEK certificate is stored.
```
```suggestion
    The user needs to provide a path to the VCEK certificate.
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+
+    Verifies the trusted compute base using the provided attestation report and VCEK certificate
+    The user needs to provide a path to the attestation report and to the directory where the VCEK certificate is stored.
+    If no path is provided for the attestation report, the program will look for it in the local directory ("./attestation_report.bin").
```
```suggestion
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    Verifies the trusted compute base using the provided attestation report and VCEK certificate
+    The user needs to provide a path to the attestation report and to the directory where the VCEK certificate is stored.
+    If no path is provided for the attestation report, the program will look for it in the local directory ("./attestation_report.bin").
+    If no path is provided for the certificate, the program will look for it in the local cert directory ("./certs").
```
```suggestion
```

Just require it altogether.

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    options:
+    -h, --help show a help message
+
+*snpguest verify attestation-signature*::
```
```suggestion
*snpguest verify signature*::
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    -h, --help show a help message
+
+*snpguest verify attestation-signature*::
+    usage: snpguest verify attestation-signature [-c, --certs] $CERTS_PATH [-a, --att-report] $ATTESTATION_REPORT_PATH
```
```suggestion
    usage: snpguest verify signature $CERTS_PATH
```

Fetch the attestation report

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    usage: snpguest verify attestation-signature [-c, --certs] $CERTS_PATH [-a, --att-report] $ATTESTATION_REPORT_PATH
+
+    Verifies that the provided attestation report was signed by the VCEK.
+    The user needs to provide a path to the attestation report and to the VCEK certificate.
+    If no path is provided for the attestation report, the program will look for it in the local directory ("./attestation_report.bin").
+    If no path is provided for the certificate, the program will look for it in the local cert directory ("./certs").
```
```suggestion
```

tylerfanelli commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    -h, --help show a help message
+
+*snpguest guest display report*::
+    usage: snpguest display report [-a, --att-report] $ATTESTATION_REPORT_PATH
```
Fetch the report?

tylerfanelli reviewed (changes requested):
Thanks @DGonzalezVillal , overall looks really good.

In a lot of these commands, we're relying on serialized files for attestation reports to be passed. In my opinion this isn't necessary, since we have access to `/dev/sev-guest` and can fetch it from the PSP ourselves. Why not just fetch the report when we can?

larrydewey commented on Cargo.toml:
```diff
@@ -14,3 +14,16 @@ exclude = [ ".gitignore", ".github/*" ]
+structopt = "0.3"
+env_logger = "0.10.0"
+anyhow = "1.0.69"
+sev = { git = "https://github.com/virtee/sev", branch = "main", features = ['openssl']}
```
+1 for this

larrydewey commented on docs/snpguest.1.adoc:
```diff
@@ -25,7 +25,95 @@ GLOBAL OPTIONS
+    Requests an attestation report from the host. Will write the contents of the attestation report in binary format into the specified report path.
+    If no path is provided for the attestation report, the path will default to current directory ("./attestation_report.bin").
+    User can pass 64 bytes of data in any file format into request in order to use that data to request the attestation report.
+    The user can use the --random flag to generate a request-file with randomly generated 64 byts. If the user provides a request path and uses the --random flag, the data will be written into the provided file path.
```
+1 For this

tylerfanelli reviewed (approved):
Nice work! We can cut a release now.

dubek commented on src/certs/mod.rs:
```diff
@@ -0,0 +1,101 @@
+// Function to check if certificate is in .der or .pem depending on its contents
+pub fn identify_cert(buf: &[u8]) -> CertFormat {
+    // Pem certificates will start with this byte content
+    const PEM_START: &[u8] = &[
+        45, 45, 45, 45, 45, 66, 69, 71, 73, 78, 32, 67, 69, 82, 84, 73, 70, 73, 67, 65, 84, 69, 45,
+        45, 45, 45, 45,
+    ];
```
maybe clearer:

```rust
const PEM_START: &[u8] = b"-----BEGIN CERTIFICATE-----";
```

(I think this is identical.)
